In [ ]:
Github Repository URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
Github commit ID: 85f23f2

In [37]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [38]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data
print(dataset.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(dataset.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [39]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset[['Temperature °C', 'Mols KCL']] # Features
Y = dataset['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)


## Part 3. Perform a Linear Regression

In [40]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, Y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = [[427, 593]]
predicted_output = model.predict(sample_data)
print(f"Predicted output for the sample data {sample_data}: {predicted_output[0]}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"coefficients: {coefficients}")
print(f"intercept: {intercept}")

Predicted output for the sample data [[427, 593]]: 572841.2133524799
coefficients: [ 866.14641337 1032.69506649]
intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


### Equation for h(x):

$$
h(x) = 866.14641 \cdot \text{Temperature °C} - 1032.69506 \cdot \text{Mols KCL} + (-409391.47958)
$$

### What Does the Score Mean?

The score R^2 represents the variance among the dependent variable of the dataset, which in this case is "Size nm^3", that is explained by the indipendent variables "Temperature °C" and "Mols KCL".

A score closer to 1 indicates that the model explains most of the variance in the dataset, or that the model fits well. Vice versa applies when the score is close to 0 or negative. 

## Part 4. Use Cross Validation

In [41]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, Y, cv=5, scoring='r2') 

print(f"Cross-validation R^2 scores: {scores}")
print(f"Mean R^2 score: {scores.mean()}")
print(f"Standard deviation of R^2 scores: {scores.std()}")

Cross-validation R^2 scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean R^2 score: 0.8568167899144437
Standard deviation of R^2 scores: 0.01346630737209602


### Findings & Significance:

The cross validation scores are consistent with each other across the board, with a range from 0.839-0.872. Since the R^2 value is near 1, we can assume the model fits the data well. The mean score of R^2 is a representation of the average performance of the model across all scores, meaning that on average, the model explains 85.7% variance in the target variable. The standard deviation of the R^2 scores is very small, about 0.013, meaning that the performance is consistent across all splits of data. This means the model generalizes very well.

## Part 5. Using Polynomial Regression

In [42]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Train the model
poly_model = LinearRegression()
poly_model.fit(X_poly, Y)

# Evaulate model w/ transformed dataset
poly_scores = cross_val_score(poly_model, X_poly, Y, cv=5, scoring='r2')

print(f"Polynomial Regression R² scores: {poly_scores}")
print(f"Mean R² score: {poly_scores.mean()}")
print(f"Standard deviation of R² scores: {poly_scores.std()}")

# Make predictions using the polynomial regression model
sample_data_poly = poly.transform([[427, 593]])  # Use the same sample point but transform it to polynomial features
predicted_output_poly = poly_model.predict(sample_data_poly)
print(f"Predicted output for the sample data (polynomial regression): {predicted_output_poly[0]}")

# Extract coefficients and intercept
coefficients_poly = poly_model.coef_
intercept_poly = poly_model.intercept_

print(f"Polynomial coefficients: {coefficients_poly}")
print(f"Polynomial intercept: {intercept_poly}")

Polynomial Regression R² scores: [1. 1. 1. 1. 1.]
Mean R² score: 1.0
Standard deviation of R² scores: 0.0
Predicted output for the sample data (polynomial regression): 521593.11427858996
Polynomial coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23110504e-07 -1.05648823e-11
  2.00000000e+00  2.85714287e-02]
Polynomial intercept: 1.657189568504691e-05


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


### Equation for h(x):
$$
h(x) = 1.65719 \times 10^{-5} + 12.0 \cdot (\text{Temperature °C}) - 1.23111 \times 10^{-7} \cdot (\text{Mols KCL}) - 1.05649 \times 10^{-11} \cdot (\text{Temperature °C})^2 + 2.0 \cdot (\text{Mols KCL})^2 \\
+ 2.85714 \times 10^{-2} \cdot (\text{Temperature °C} \cdot \text{Mols KCL})
$$
### Metrics Report:
For the data sample [427, 593], the predicted output from the polynomial regression model is 521,841.21 which is similar to the linear regression model number of 572,841.21. The cross validation scores for polynomial regression are impressive, as they indicate that the polynomial regression model perfectly fits the training data across all folds with a R^2 score of 1 across all folds, which means it explains 100% of the variance in the target variable. This could indicate that the polynomial model is overfitting the dataset. 